packages/server/postgres/migrations/1709478683000_moveInvoice.ts

```diff
@@ -0,0 +1,88 @@
+import {FirstParam} from 'parabol-client/types/generics'
+import {Client} from 'pg'
+import {r} from 'rethinkdb-ts'
+import getPgConfig from '../getPgConfig'
+import connectRethinkDB from '../../database/connectRethinkDB'
+import getPgp from '../getPgp'
+
+export async function up() {
+  await connectRethinkDB()
+  const {pgp, pg} = getPgp()
+  const batchSize = 1000
+
+  try {
+    await r.table('Invoice').indexCreate('startAt').run()
+    await r.table('Invoice').indexWait().run()
+  } catch {}
+
+  const columnSet = new pgp.helpers.ColumnSet(
+    [
+      'amountDue',
+      'createdAt',
+      {name: 'coupon', def: null},
+      'total',
+      'billingLeaderEmails',
+      {name: 'creditCard', def: null},
+      'endAt',
+      'invoiceDate',
+      {name: 'lines', cast: 'jsonb[]'},
+      'nextPeriodCharges',
+      'orgId',
+      'orgName',
+      {name: 'payUrl', def: null},
+      {name: 'picture', def: null},
+      'startAt',
+      'startingBalance',
+      'status',
+      'tier'
+    ],
+    {table: 'Invoice'}
+  )
+
+  const getNextData = async (leftBoundCursor: Date | undefined) => {
+    const startAt = leftBoundCursor || r.minval
+    const nextBatch = await r
+      .table('Invoice')
+      .between(startAt, r.maxval, {index: 'startAt', leftBound: 'open'})
+      .orderBy({index: 'startAt'})
+      .limit(batchSize)
+      .run()
+    if (nextBatch.length === 0) return null
+    if (nextBatch.length < batchSize) return nextBatch
+    const lastItem = nextBatch.pop()
+    const lastMatchingStartAt = nextBatch.findLastIndex(
+      (item) => item.startAt !== lastItem!.startAt
+    )
+    if (lastMatchingStartAt === -1) {
+      throw new Error(
+        'batchSize is smaller than the number of items that share the same cursor. Increase batchSize'
+      )
+    }
+    return nextBatch.slice(0, lastMatchingStartAt)
```

Comment:
Found a tricky off-by-one bug here. Turned up after I migrated the `Invoice` data, I noticed I was missing exactly 1 row.

I wrote quick code to export both tables to CSV and then wrote a quick Python script to compare them. Yielded the missing row:

```
grep Hk9-U-VcZ *.csv
pgdata.csv:"2017-09-11 12:48:02+00","Hk9-U-VcZ"
rethinkdata.csv:"2017-09-11T12:48:02.000Z","Hk9-U-VcZ"
rethinkdata.csv:"2017-09-11T12:48:02.000Z","Hk9-U-VcZ"
```

Created a conditional breakpoint in this function to find when this org was being processed and it helped me see that this function's logic works by using the last index key as the left open interval...we were inadvertently only every going to get 1 row for any duplicate key that fit within `batchSize`

Reply:
I did some testing and turns out you're right, shame on me.